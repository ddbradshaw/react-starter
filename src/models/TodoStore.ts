import { types, getParent, destroy, flow, getSnapshot, onSnapshot, onPatch, applyPatch } from "mobx-state-tree"

//TODO: Implement more modularized solution
// Patch based time travelling
// This is experimental
const patches:any[] = [];
const redoPatches:any[] = [];
let isTimeTravel:boolean = false

export const TodoItem = types
  .model("Todo", {
    userId: types.optional(types.number, 1),
    id: types.number,
    title: types.string,
    completed: types.optional(types.boolean, false)
  })
  .actions(self => {
    const ax = {
      changeLabel(newLabel: string) {
        self.title = newLabel;
      },
      changeComplete(newValue: boolean) {
        self.completed = newValue;
      },
      remove() {
        getParent(self, 2).remove(self);
      },
      save: flow(function* save() {
        try {
          yield window.fetch(`https://jsonplaceholder.typicode.com/todos/1`, {
            method: "PUT",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify(getSnapshot(self))
          })
        } catch (ex) {
          console.warn("Unable to save todo", ex)
        }
      }),
      // Using state lifecyle hooks to listen to changes to todo and auto saving to server
      // This is a contrived example.
      afterCreate() {
        onSnapshot(self, ax.save)
      }
    }
    return ax;
  })

export const TodoStore = types
  .model("TodoStore", {
    todos: types.optional(types.array(TodoItem), []),
    loading: types.optional(types.boolean, false)
  })
  .actions(self => ({
    add(todo: ITodoItem) {
      self.todos.push(todo);
    },
    remove(todo: ITodoItem) {
      destroy(todo);
    },
    undo() {
      if(patches.length == 0 ) return;
      isTimeTravel = true;
      const p = patches[patches.length - 1]; // Get the last patch in the heap
      redoPatches.push(p); // Save the patch to the redo heap
      applyPatch(self, p.inversePatch); // Apply the patch inverse
      patches.splice(patches.length - 2, 2); // Remove the reversed actions from heap
    },
    redo() {
      if(redoPatches.length == 0 ) return;
      isTimeTravel = true;
      const p = redoPatches[redoPatches.length - 1]; // Get the last patch in the heap
      applyPatch(self, p.patch); // Apply the patch inverse
      redoPatches.splice(redoPatches.length - 1, 1); // Remove the reversed actions from heap

    },
    getSuggestions: flow(function* getSuggestions(count: number) {
      self.loading = true;
      try {
        const response = yield window.fetch('https://jsonplaceholder.typicode.com/todos');
        const todos = yield response.json();
        const slice = todos.slice(0, count); // Grab top 5 (since service return 200)
        slice.map((item: ITodoItem) => { // Randomize the ID
          item.id = Math.floor(Math.random() * 100000000)
        })
        self.todos.push(...slice);
        self.loading = false;
      } catch (ex) {
        console.error(ex);
        self.loading = false;
      }
    }),
    // Listen for patches to the model.  Save the patches to the heap for undo/redo
    afterCreate() {
      onPatch(self, (patch, inversePatch) => {
        patches.push({patch, inversePatch});
        if(!isTimeTravel) redoPatches.splice(0, redoPatches.length);
        isTimeTravel = false;
      })
    }
  }))
  // view functions are extracted into functions so that 
  // we can use view functions within other view functions
  .views(self => {
    const vx = {
      get todoCount() {
        return self.todos.length;
      },
      get completedCount() {
        return self.todos.filter(todo => todo.completed == true).length;
      },
      get completedPct() {
        if (vx.todoCount == 0) return 0;
        return Math.floor((vx.completedCount / vx.todoCount) * 100);
      }
    }
    return vx;
  })

export type ITodoItem = typeof TodoItem.Type;
export type ITodoStore = typeof TodoStore.Type;