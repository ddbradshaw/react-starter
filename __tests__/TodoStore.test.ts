import { getSnapshot, onSnapshot, onPatch } from 'mobx-state-tree'
import { TodoItem, TodoStore, ITodoItem, ITodoStore } from '../src/models/TodoStore'

it("can create an instance of a model", () => {
  const todo = TodoItem.create({
    id: 1,
    label: "Test Todo",
    complete: false
  });

  expect(todo.label).toBe("Test Todo");
  expect(todo.complete).toBe(false);
  todo.changeLabel("New Label")
  expect(todo.label).toBe("New Label");
})

it("can create a new todo store", () => {
  const store = TodoStore.create({
    todos: [{
      id: 1,
      label: "Test Todo",
      complete: false
    }]
  })

  expect(store.todos.length).toBe(1);
  expect(store.todos[0].label).toBe("Test Todo");
})

it("can add new todos - track snapshots", () => {
  const store = TodoStore.create();
  const states: Array<any> = [];

  onSnapshot(store, snapshot => {
    states.push(snapshot)
  })

  store.add({
    id: 1,
    label: "Test Todo",
    complete: false
  } as ITodoItem)

  expect(store.todos.length).toBe(1);
  expect(store.todos[0].label).toBe("Test Todo");
  expect(getSnapshot(store)).toEqual({
    todos: [{
      id: 1,
      label: "Test Todo",
      complete: false
    }]
  })
  expect(getSnapshot(store)).toMatchSnapshot();
  expect(states).toMatchSnapshot();
})

it("can add new items - track patches", () => {
  const store = TodoStore.create();
  const patches: Array<any> = [];

  onPatch(store, patch => {
    patches.push(patch)
  })

  store.add({
    id: 1,
    label: "Test Todo",
    complete: false
  } as ITodoItem)

  store.todos[0].changeLabel("New Label");
  expect(patches).toMatchSnapshot();
})

it("can calculate the total length", () => {
  const store = TodoStore.create({
    todos: [{
      id: 1,
      label: "Test Todo",
      complete: false
    }]
  })
  expect(store.todoCount).toBe(1);
})