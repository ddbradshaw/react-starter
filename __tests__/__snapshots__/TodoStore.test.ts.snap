// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add new items - track patches 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/todos/0",
    "value": Object {
      "complete": false,
      "id": 1,
      "label": "Test Todo",
    },
  },
  Object {
    "op": "replace",
    "path": "/todos/0/label",
    "value": "New Label",
  },
]
`;

exports[`can add new todos - track snapshots 1`] = `
Object {
  "todos": Array [
    Object {
      "complete": false,
      "id": 1,
      "label": "Test Todo",
    },
  ],
}
`;

exports[`can add new todos - track snapshots 2`] = `
Array [
  Object {
    "todos": Array [
      Object {
        "complete": false,
        "id": 1,
        "label": "Test Todo",
      },
    ],
  },
]
`;
